htmd/parameterization/dihedral.py

```diff
@@ -0,0 +1,548 @@
+
+        types = [self.molecule._rtf.type_by_index[index] for index in dihedral]
+
+        all_uses = []
+        for dihedral_indices in self.molecule.dihedrals:
+            dihedral_types = [self.molecule._rtf.type_by_index[index] for index in dihedral_indices]
+            if types == dihedral_types or types == dihedral_types[::-1]:
+                all_uses.append(dihedral_indices)
+
+        # Now for each of the uses, remove any which are equivalent
+        unique_uses = [dihedral]
+        groups = [self.molecule._equivalent_group_by_atom[index] for index in dihedral]
+        for dihed in all_uses:
+            dihedral_groups = [self.molecule._equivalent_group_by_atom[index] for index in dihed]
+            if groups != dihedral_groups and groups != dihedral_groups[::-1]:
+                unique_uses.append(dihed)
+
+        return unique_uses
+
+    def _makeDihedralUnique(self, dihedral):
+        """
+        Duplicate atom types of the dihedral, so its parameters are unique.
+        """
+        # TODO check symmetry
+
+        # Duplicate the atom types of the dihedral
+        for i in range(4):
+            if not ("x" in self.molecule._rtf.type_by_index[dihedral[i]]):
+                self.molecule.duplicateTypeOfAtom(dihedral[i])
+
+        equivalent_dihedrals = self._getEquivalentDihedrals(dihedral)
+        if len(equivalent_dihedrals) > 1:
+            print(dihedral)
+            print(len(equivalent_dihedrals))
+            print(equivalent_dihedrals)
+            raise ValueError("Dihedral term still not unique after duplication")
+
+    def _getValidQMResults(self):
+        """
+        Get a set of valid QM results
+        """
+
+        all_valid_results = []
+        for results in self.qm_results:
+
+            # Remove failed QM results
+            # TODO print removed QM jobs
+            valid_results = [result for result in results if not result.errored]
+
+            # Remove QM results with too high QM energies (>20 kcal/mol above the minimum)
+            # TODO print removed QM jobs
+            qm_min = np.min([result.energy for result in valid_results])
+            valid_results = [result for result in valid_results if (result.energy - qm_min) < 20]
+
+            if len(valid_results) < 13:
+                raise RuntimeError('Fewer than 13 valid QM points. Not enough to fit!')
+
+            all_valid_results.append(valid_results)
+
+        return all_valid_results
+
+    def _setup(self):
+
+        if len(self.dihedrals) != len(self.qm_results):
+            raise ValueError('The number of dihedral and QM result sets has to be the same!')
+
+        # Get dihedral names
+        self._names = ['%s-%s-%s-%s' % tuple(self.molecule.name[dihedral]) for dihedral in self.dihedrals]
+
+        # Get equivalent dihedral atom indices
+        self._equivalent_indices = []
+        for idihed, dihedral in enumerate(self.dihedrals):
+            for rotableDihedral in self.molecule._soft_dihedrals:
```

Comment:
Let's try to make nomenclature simple: from what I know, soft and rotatable torsions are equivalent terms in the parameterize context. Just use one.

Reply:
I'm uniforming names, but this comes from not yet re-factored part.